Refactor anti-entropy and local state

magiconair opened the pull request:
This patch decouples the local state and the anti-entropy code from the agent and from each other to have cleaner separation of concerns and to ensure that data structures are properly locked.

preetapan commented on agent/ae/ae.go:
```diff
@@ -93,67 +93,64 @@ func (s *StateSyncer) Run() {
 
 FullSync:
```
This looks right, but also looks tricky to get right. Is there a unit test that exercises this part of the code already that passes before and after?

magiconair replied:
Need to check.

magiconair replied:
I think so since I've moved the tests as well but I'll check. Not sure how much they cover.

preetapan commented on agent/ae/ae.go:
```diff
@@ -47,100 +55,132 @@ type StateSyncer struct {
+
+// Pause temporarily disables partial sync runs.
+func (s *StateSyncer) Pause() {
+	atomic.AddInt32(&s.paused, 1)
```
Curious why pause adds instead of setting to one. Do we expect multiple calls to Pause before Resume? 

Not something you changed, looks like it was already working like that, but I'd like to understand why

magiconair replied:
I think this is defensive programming but I agree that this looks like overkill.

magiconair replied:
I agree that this looks like belt and suspenders. I'll have another look whether this is really necessary.

preetapan commented on agent/ae/ae.go:
```diff
@@ -47,100 +55,132 @@ type StateSyncer struct {
-	// Schedule the next full sync, with a random stagger
-	aeIntv := aeScale(ae.Interval, ae.ClusterSize())
-	aeIntv = aeIntv + lib.RandomStagger(aeIntv)
-	aeTimer := time.After(aeIntv)
```
why did this timer stop being useful

magiconair replied:
This is still there. I've just inlined it.

magiconair commented:
I've rebased the branch and fixed a conceptual mistake I made. I've renamed the `agent.setSyncState` method to `SyncFull` since I've somehow come to the conclusion that there was a full and partial sync of state in the agent. However, the `setSyncState` flags the state which is out-of-sync and `syncChanges` then performs the actual sync. The changes I've force-pushed reflect that now.

slackpad commented on agent/ae/ae.go:
```diff
@@ -0,0 +1,184 @@
+type StateSyncer struct {
+	// paused flags whether sync runs are temporarily disabled.
+	// Must be the first element due to a go bug.
+	// todo(fs): which bug? Is this still relevant?
```
https://golang.org/pkg/sync/atomic/#pkg-note-BUG

magiconair replied:
The comment says this:

> On both ARM and x86-32, it is the caller's responsibility to arrange for 64-bit alignment of 64-bit words accessed atomically. The first word in a variable or in an allocated struct, array, or slice can be relied upon to be 64-bit aligned.

Since `paused` is an `int32` I don't think this applies. However, this is too easy to miss and I therefore suggest to refactor this as follows:

```go
// toggle implements an on/off switch using methods from the atomic
// package. Since fields in structs that are accessed via
// atomic.Load/Add methods need to be aligned properly on some platforms
// we move that code into a separate struct.
//
// See https://golang.org/pkg/sync/atomic/#pkg-note-BUG for details
type toggle int32

func (p toggle) On() {
	atomic.AddInt32(&p, 1)
}

func (p toggle) Off() {
	if atomic.AddInt32(&p, -1) < 0 {
		panic("toggle not on")
	}
}

func (p toggle) IsOn() bool {
	return atomic.LoadInt32(&p) > 0
}
```

slackpad commented on agent/ae/ae.go:
```diff
@@ -0,0 +1,184 @@
+
+FullSync:
+	for {
+		switch err := s.State.SyncFull(); {
```
Why a `switch` here instead of an `if`/`else` (esp. with the nested `switch`?

magiconair replied:
Personal preference and leftover from previous refactor attempts. You're right that an `if/else` de-clutters this.

slackpad commented on agent/ae/ae.go:
```diff
@@ -0,0 +1,184 @@
+					continue FullSync
+
+				case <-s.TriggerCh:
+					if s.Paused() {
```
Should we respect `Paused()` up above for the full sync as well?

magiconair replied:
If we do this then we should timebox the pausing since otherwise you may end up in a state where you never to a full sync again. When are we pausing syncs right now?

slackpad commented on agent/ae/ae.go:
```diff
@@ -0,0 +1,184 @@
+			// do partial syncs until it is time for a full sync again
+			for {
+				select {
+				// todo(fs): why don't we honor the ServerUpCh here as well?
```
Maybe, might be a good place to `continue FullSync` so that things converge better from here as well.

slackpad commented on agent/agent.go:
```diff
@@ -237,6 +243,21 @@ func New(c *Config) (*Agent, error) {
 	return a, nil
 }
 
+func LocalConfig(cfg *Config) local.Config {
```
Note to self - double check how node metadata is handled - it's not here.

slackpad commented on agent/agent.go:
```diff
@@ -254,17 +275,44 @@ func (a *Agent) Start() error {
 		return fmt.Errorf("Failed to setup node ID: %v", err)
 	}
 
+	// create a notif channel to trigger state sychronizations
```
"notif" here and on line 282

slackpad commented on agent/agent.go:
```diff
@@ -254,17 +275,44 @@ func (a *Agent) Start() error {
 
+	// create a notif channel to trigger state sychronizations
+	// when a consul server was added to the cluster.
+	serverUpCh := make(chan struct{}, 1)
```
I'd move this down since the `local` doesn't need it.

slackpad commented on agent/agent.go:
```diff
@@ -254,17 +275,44 @@ func (a *Agent) Start() error {
+	// ServerUp is used to inform that a new consul server is now
+	// up. This can be used to speed up the sync process if we are blocking
+	// waiting to discover a consul server
+	// todo(fs): IMO, the non-blocking nature of this call should be hidden in the syncer
```
That's not a bad idea to just call `local.NotifyNewServer()` instead of plumbing a channel here.

magiconair replied:
The problem is that you need the non-blocking nature in two places: in the local state and the AE code. If we want to hide this we have two options: callback or move the nonblocking feature into a separate struct. The latter feels like gold-plating whereas the former works against the decoupling. In essence, we want a channel which is not blocking on write.

magiconair replied:
I've added an `ae.Trigger` which hides this implementation detail and wire them up with callbacks between the state, syncer and the consul rpc server.

slackpad commented on agent/agent.go:
```diff
@@ -274,15 +322,17 @@ func (a *Agent) Start() error {
 		a.delegate = server
-		a.state.delegate = server
+		a.State.SetDelegate(server)
+		a.sync.ClusterSize = func() int { return len(server.LANMembers()) }
```
How about we do this once after this using the delegate? `a.sync.ClusterSize = func() int ( return len(a.delegate.LANMembers()) }`

magiconair replied:
yes

magiconair replied:
done.

slackpad commented on agent/agent.go:
```diff
@@ -1140,18 +1190,18 @@ func (a *Agent) WANMembers() []serf.Member {
-	// Start the anti entropy routine
-	go a.state.antiEntropy(a.shutdownCh)
+	go a.sync.Run()
+	a.logger.Printf("[INFO] agent: starting state syncer")
```
"started"

magiconair replied:
done

slackpad commented on agent/agent.go:
```diff
@@ -1211,29 +1261,31 @@ func (a *Agent) sendCoordinate() {
 		}
 
 		// See if there's a timeout.
+		// todo(fs): this looks fishy... why is there anoter data structure in the agent with its own lock?
```
Yeah this was kind of a hack since we had those awful parallel maps for each check type - this is logically a member of a check object.

magiconair replied:
ok, I'll address that in a separate PR

slackpad commented on agent/agent.go:
```diff
@@ -1982,12 +2032,11 @@ func (a *Agent) loadServices(conf *Config) error {
 // unloadServices will deregister all services other than the 'consul' service
```
This comment is stale (the `consul` service is no longer special).

magiconair replied:
done

slackpad commented on agent/agent.go:
```diff
@@ -2070,52 +2119,39 @@ func (a *Agent) loadChecks(conf *Config) error {
-	for _, check := range a.state.Checks() {
-		if err := a.RemoveCheck(check.CheckID, false); err != nil {
-			return fmt.Errorf("Failed deregistering check '%s': %s", check.CheckID, err)
+	for id := range a.State.Checks() {
```
I wonder if the unload for services and checks should be added as a method on the state, since this is technically working off a copy so this isn't atomic.

slackpad commented on agent/catalog_endpoint_test.go:
```diff
@@ -13,6 +13,8 @@ import (
 )
 
 func TestCatalogRegister(t *testing.T) {
+	t.Skip("skipping since it is not clear what this test is supposed to verify")
```
I can't think of a good way to make this useful - let's just delete this test.

magiconair replied:
done

slackpad commented on agent/local/state.go:
```diff
@@ -0,0 +1,972 @@
+	"github.com/hashicorp/consul/types"
+)
+
+// permissionDenied is returned when an ACL based rejection happens.
```
This can be deleted - the `acl` package has the standard form of this (and I don't think this is referenced).

magiconair replied:
done

slackpad commented on agent/local/state.go:
```diff
@@ -0,0 +1,972 @@
+	// Service is the local copy of the service record.
+	Service *structs.NodeService
+
+	// Token is the ACL to update the service record on the server.
```
I'd say "update or delete" since that's kind of a subtle aspect of this.

magiconair replied:
done

slackpad commented on agent/local/state.go:
```diff
@@ -0,0 +1,972 @@
+	// Check is the local copy of the health check record.
+	Check *structs.HealthCheck
+
+	// Token is the ACL record to update the health check record
```
"update or delete"

magiconair replied:
done

slackpad commented on agent/local/state.go:
```diff
@@ -0,0 +1,972 @@
+	CriticalTime time.Time
+
+	// DeferCheck is used to delay the sync of a health check when
+	// only the status has changed.
```
It's actually for when the status _has not_ changed but there has been a change to the output. It's used to rate limit changes like that which don't affect the state of the node/service since the status didn't update.

magiconair replied:
done

magiconair commented:
@slackpad I've addressed most review comments except for moving the two methods into the local state. I'll think about that. 

> Where do you envision moving the maps with the actual checks and the check runners down the road?

I'd like to move them into a separate struct which manages checks and services instead of the multiple maps in the agent.

slackpad commented on agent/local/state.go:
```diff
@@ -0,0 +1,963 @@
+	var out struct{}
+	err := l.Delegate.RPC("Catalog.Deregister", &req, &out)
+	if err == nil || strings.Contains(err.Error(), "Unknown check") {
+		// todo(fs): do we need to stop the deferCheck timer here?
```
It reloads the check and skips out if the check is gone, so likely ok as-is, but stopping it is probably nice to do.

magiconair replied:
stopping the timer now

magiconair commented on agent/local/state.go:
```diff
@@ -0,0 +1,963 @@
+// and checks. We used it to perform anti-entropy with the
+// catalog representation
+type State struct {
+	sync.RWMutex
```
Yeah, didn't get to that yet. First remove all external usage of the lock, then make it private.

magiconair commented on agent/local/state.go:
```diff
@@ -0,0 +1,963 @@
+	// we are calling will do that themself.
+
+	// todo(fs): is it an issue that we do not hold the lock for the entire time?
+	// todo(fs): IMO, this doesn't matter since SyncChanges will sync whatever
```
Updated the comment

magiconair commented on agent/local/state.go:
```diff
@@ -0,0 +1,963 @@
+		}
+
+		// If the service is scheduled for removal skip it.
+		// todo(fs): is this correct?
```
ok

magiconair commented on agent/local/state.go:
```diff
@@ -0,0 +1,963 @@
+	// Checks tracks the local checks
+	checks map[types.CheckID]*CheckState
+
+	// metadata tracks the local metadata fields
```
done

magiconair commented on agent/local/state.go:
```diff
@@ -0,0 +1,963 @@
+		return nil
+	}
+	if acl.IsErrPermissionDenied(err) {
+		// todo(fs): why is the service in sync here?
```
added a comment.

magiconair commented on agent/ae/ae.go:
```diff
@@ -0,0 +1,215 @@
+			case <-time.After(s.Interval + stagger(s.Interval)):
+				continue FullSync
+
+			// do patial syncs on demand
```
done

magiconair commented on agent/ae/ae.go:
```diff
@@ -0,0 +1,215 @@
+FullSync:
+	for {
+		// attempt a full sync
+		if err := s.State.SyncFull(); err != nil {
```
The challenge here is not to get into a busy wait loop which would happen when you return `nil`. I'm not too fond of the two loops but turning this into a state machine didn't make this prettier.

How about this:

```go
// Run is the long running method to perform state synchronization
// between local and remote servers.
func (s *StateSyncer) Run() {
	if s.ClusterSize == nil {
		panic("ClusterSize not set")
	}

	stagger := func(d time.Duration) time.Duration {
		f := scaleFactor(s.ClusterSize())
		return lib.RandomStagger(time.Duration(f) * d)
	}

FullSync:
	for {
		// attempt a full sync
		err := s.ifNotPausedRun(s.State.SyncFull)
		if err != nil {
			if err != errPaused {
				s.Logger.Printf("[ERR] agent: failed to sync remote state: %v", err)
			}

			select {
			// trigger a full sync immediately.
			// this is usually called when a consul server was added to the cluster.
			// stagger the delay to avoid a thundering herd.
			case <-s.SyncFull.Notif():
				select {
				case <-time.After(stagger(serverUpIntv)):
				case <-s.ShutdownCh:
					return
				}

			// retry full sync after some time
			// todo(fs): why don't we use s.Interval here?
			case <-time.After(retryFailIntv + stagger(retryFailIntv)):

			case <-s.ShutdownCh:
				return
			}

			continue
		}

		// do partial syncs until it is time for a full sync again
		for {
			select {
			// trigger a full sync immediately
			// this is usually called when a consul server was added to the cluster.
			// stagger the delay to avoid a thundering herd.
			case <-s.SyncFull.Notif():
				select {
				case <-time.After(stagger(serverUpIntv)):
					continue FullSync
				case <-s.ShutdownCh:
					return
				}

			// time for a full sync again
			case <-time.After(s.Interval + stagger(s.Interval)):
				continue FullSync

			// do partial syncs on demand
			case <-s.SyncChanges.Notif():
				err := s.ifNotPausedRun(s.State.SyncChanges)
				if err != nil && err != errPaused {
					s.Logger.Printf("[ERR] agent: failed to sync changes: %v", err)
				}

			case <-s.ShutdownCh:
				return
			}
		}
	}
}

func (s *StateSyncer) ifNotPausedRun(f func() error) error {
	s.pauseLock.Lock()
	defer s.pauseLock.Unlock()
	if s.paused {
		return errPaused
	}
	return f()
}

// Pause temporarily disables sync runs.
func (s *StateSyncer) Pause() {
	s.pauseLock.Lock()
	if s.paused {
		panic("pause while paused")
	}
	s.paused = true
	s.pauseLock.Unlock()
}

// Paused returns whether sync runs are temporarily disabled.
func (s *StateSyncer) Paused() bool {
	s.pauseLock.Lock()
	defer s.pauseLock.Unlock()
	return s.paused
}

// Resume re-enables sync runs.
func (s *StateSyncer) Resume() {
	s.pauseLock.Lock()
	if !s.paused {
		panic("resume while not paused")
	}
	s.paused = false
	s.pauseLock.Unlock()
	s.SyncChanges.Trigger()
}
```

magiconair replied:
another option to make things a bit more explicit:

```go
// Run is the long running method to perform state synchronization
// between local and remote servers.
func (s *StateSyncer) Run() {
	if s.ClusterSize == nil {
		panic("ClusterSize not set")
	}

	stagger := func(d time.Duration) time.Duration {
		f := scaleFactor(s.ClusterSize())
		return lib.RandomStagger(time.Duration(f) * d)
	}

FullSync:
	for {
		// attempt a full sync
		err := s.ifNotPausedRun(s.State.SyncFull)
		if err == nil {
			goto PartialSync
		}

		if err != errPaused {
			s.Logger.Printf("[ERR] agent: failed to sync remote state: %v", err)
		}

		select {
		// trigger a full sync immediately; usually when a consul server was added
		// stagger to avoid a thundering herd.
		case <-s.SyncFull.Notif():
			select {
			case <-time.After(stagger(serverUpIntv)):
				continue FullSync
			case <-s.ShutdownCh:
				return
			}

		case <-time.After(retryFailIntv + stagger(retryFailIntv)):
			continue FullSync

		case <-s.ShutdownCh:
			return
		}

	PartialSync:
		for {
			select {
			// trigger a full sync immediately; usually when a consul server was added
			// stagger to avoid a thundering herd.
			case <-s.SyncFull.Notif():
				select {
				case <-time.After(stagger(serverUpIntv)):
					continue FullSync
				case <-s.ShutdownCh:
					return
				}

			// time for a full sync again
			case <-time.After(s.Interval + stagger(s.Interval)):
				continue FullSync

			// do partial syncs on demand
			case <-s.SyncChanges.Notif():
				err := s.ifNotPausedRun(s.State.SyncChanges)
				if err != nil && err != errPaused {
					s.Logger.Printf("[ERR] agent: failed to sync changes: %v", err)
				}
				continue PartialSync

			case <-s.ShutdownCh:
				return
			}
		}
	}
}
```

slackpad commented on agent/local/state.go:
```diff
@@ -0,0 +1,1031 @@
+		if s.Deleted {
+			continue
+		}
+		m[id] = s.Service
```
Should this be a `Clone()`?

slackpad replied:
I suppose not since this is the immutable part.

magiconair replied:
We only do `Clone` for `ServiceState` and `CheckState`.

magiconair replied:
Also, `Clone` is shallow in that it will still return the pointer to the `NodeService`. See comment in line 50.

slackpad commented on agent/local/state.go:
```diff
@@ -0,0 +1,1031 @@
+	}
+
+	if l.discardCheckOutput.Load().(bool) {
+		check.Output = ""
```
This feels like an unexpected side effect since you are passing in a pointer.

magiconair replied:
If we wouldn't do this here then the first check might be stored with output but subsequent checks have their output discarded. In this case you'd update the server once even when the status hasn't changed and `discard_check_output=true`. Granted, this is in all likelihood only an issue during testing since only there you'll register a check with some initial output but that's how I stumbled over this.

I agree that this is a side effect but it is in one place (`AddCheck` and `UpdateCheck`). Otherwise, I'd have to move the discarding to the check registration in the API and the check execution.

If we do not want to modify the check record at all then we need to modify the compare and sync logic to push a clone of the record to the server which does not contain the output. I'm not sure that's worth it but I'll think about it.

magiconair replied:
fixed.

slackpad commented on agent/local/state.go:
```diff
@@ -0,0 +1,1031 @@
+
+	// if there is a serviceID associated with the check, make sure it exists before adding it
+	// NOTE - This logic may be moved to be handled within the Agent's Addcheck method after a refactor
+	if check.ServiceID != "" && l.services[check.ServiceID] == nil {
```
You need to be holding the lock in order to read `l.services` here.

magiconair replied:
Indeed

slackpad commented on agent/local/state.go:
```diff
@@ -0,0 +1,1031 @@
+	}
+
+	// hard-set the node name
+	check.Node = l.config.NodeName
```
Same here - might be slightly better to operate on a copy since we are going to modify the passed-in structure.

magiconair replied:
yep

slackpad commented on agent/local/state_test.go:
```diff
@@ -1401,100 +1261,76 @@ func TestAgentAntiEntropy_NodeInfo(t *testing.T) {
-		}
-	})
+	// Make sure we synced our node info - this should have ridden on the
+	// "consul" service sync
```
This comment is stale - we can just delete it.

magiconair commented on agent/ae/ae.go:
```diff
@@ -0,0 +1,320 @@
+	if s.paused < 0 {
+		panic("unbalanced pause/resume")
+	}
+	if s.paused == 0 {
```
Trigger won't block but I can refactor.

https://github.com/hashicorp/consul/pull/3391/files#diff-1af33268cdd38b054f30645f00b40bb0R14

magiconair commented on agent/ae/ae_test.go:
```diff
@@ -0,0 +1,397 @@
+		l := testSyncer()
+		l.ShutdownCh = make(chan struct{})
+		evch := make(chan event)
+		go func() { evch <- l.syncChangesEvent() }()
```
`syncChangesEvent` and `retryFullSyncEvent` are similar but different `select` statements. The code paths for `shutdown` and `shutdown while waiting to handle server up event` are the same but duplicated. That's why we need to test them twice.

slackpad replied:
I misread this and didn't see you consuming the event. Makes sense now.

magiconair commented:
Addressed review comments.
